BUG: don't silence warnings in ufunc.reduce

eric-wieser opened the pull request:
Needed a file of helper functions so that this was actually visible in `reduction.c`. This move is in the first commit.

Fix itself is trivial once things are visible.

Fixes #8954

eric-wieser committed:
MAINT: Move functions handling ufunc extobj

This makes it visible inside other files, like reduction.c

Apart from the includes, everything here is just a direct move - there are no code changes

eric-wieser commented on numpy/core/src/umath/extobj.c:
```diff
@@ -0,0 +1,310 @@
+#include "extobj.h"
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+#include <numpy/ufuncobject.h>
+#include "ufunc_object.h"  /* for npy_um_str_pyvals_name */
```
Is there a right order for this includes?

eric-wieser commented:
Tests fail right now because we're now being warned about more things. Most of these are correct (ie, testing nan behavior), and need test updates.

I don't think these should warn at all when invoked with `(1, np.nan)` or `.reduce([1, np.nan])`, since they are designed to explicitly handle `nan`s:

* np.minimum
* np.maximum
* np.fmin
* np.fmax

However, I don't know how to disable the warnings there.

Also, one test fails due to an overflow that was silenced! Which raises a good point - overflow in `np.sum` is something we really should have been catching

eric-wieser committed:
MAINT: Do not produce warnings in fmin/fmax functions

Previously, these would do so when encountering nans

eric-wieser commented:
Ok, all fixed I think. Probably easiest to review this commit-by-commit, since the first refactor is noisy

juliantaylor commented:
maximum/minimum should warn, nan's may lead to unexpected results. We also have the nanmax/nanmin functions for these.

eric-wieser commented:
What's the history behind having both `fmax` and `nanmax`?

eric-wieser committed:
BUG: Check for warnings in ufunc.reduce

The two most visible consequences are warnings for:
* np.min and np.max when inputs contain nan
* np.sum when the summation (of floats) overflows

Implementation taken from the code for ufunc.__call__

Fixes #8954

eric-wieser commented:
@juliantaylor: Ok, warnings are now only squashed in fmin and fmax

homu commented:
:umbrella: The latest upstream changes (presumably #8876) made this pull request unmergeable. Please resolve the merge conflicts.

homu commented:
:umbrella: The latest upstream changes (presumably #9026) made this pull request unmergeable. Please resolve the merge conflicts.

charris commented:
Needs rebase.

charris commented:
> What's the history behind having both fmax and nanmax?

`fmax` is a binary ufunc implementing the standard C fmax function. The nanmax function uses it via the reduce method when it can, but see the note in the implementation. It might can be changed now that `__array_ufunc__` is in.

```
Upon successful completion, these functions shall return the maximum numeric value of their arguments.

If just one argument is a NaN, the other argument shall be returned.
```

eric-wieser commented:
@charris: Rebased

mhvk commented on numpy/core/src/umath/extobj.c:
```diff
@@ -0,0 +1,317 @@
+    }
+
+    return 0;
+}
```
Missing end of line.

mhvk commented on numpy/core/src/umath/extobj.h:
```diff
@@ -0,0 +1,32 @@
+ufunc_update_use_defaults(void);
+#endif
+
+#endif
```
Missing end of line.

mhvk commented on numpy/core/src/umath/reduction.c:
```diff
@@ -436,6 +437,7 @@ PyArray_InitializeReduceResult(
  * loop        : The loop which does the reduction.
  * data        : Data which is passed to assign_identity and the inner loop.
  * buffersize  : Buffer size for the iterator. For the default, pass in 0.
+ * errormask   : forwarded from _get_bufsize_errmask
```
Since it comes after `funcname` in the signature, move it below it here as well.

mhvk commented on numpy/core/src/umath/reduction.h:
```diff
@@ -149,6 +149,7 @@ PyUFunc_ReduceWrapper(PyArrayObject *operand, PyArrayObject *out,
                       PyArray_ReduceLoopFunc *loop,
-                      void *data, npy_intp buffersize, const char *funcname);
+                      void *data, npy_intp buffersize, const char *funcname,
+                      int errormask);
```
Add `errormask` to the description above as well.

mhvk reviewed (commented):
Mostly nitpicks, but also warnings in the compilation:
```
    numpy/core/src/umath/ufunc_object.c:3619:15: warning: unused variable 'out_obj' [-Wunused-variable]

    numpy/core/src/umath/reduction.c:562:5: warning: implicit declaration of function 'PyUFunc_clearfperr' [-Wimplicit-function-declaration]
```
I think the first is likely a result of the rebase; the definition can be removed from l.3619, as it is moved to 3655.

The second I'm not sure about; a missing include perhaps?

eric-wieser commented:
Don't understand how this patch could have introduced the warning about `out_obj`, since the diff doesn't even contain that word, but all your nitpicks are fair

eric-wieser committed:
MAINT: Remove unused variable

eric-wieser commented:
Nitpicks addressed, but still not sure about that missing function prototype. Works fine on MSVC though...

mhvk commented:
This now looks all OK. The one remaining question is whether this is worth backporting. Including the MAINT commits, it is a bit of a large code shuffle, but it does fix a bug. @charris?

charris commented:
@mhvk That's a lot of code. If the bug is long standing I would leave it for 1.14. I'm OK with backporting fixes to `__array_ufunc__`,  but this doesn't look to qualify for that.

I'm going to release the 1.13.0 final at the beginning of next week.

eric-wieser commented:
> That's a lot of code. 

Not really - almost all of it is just moving code from one file to another. I agree though, this isn't terribly deserving of a backport

charris commented on numpy/core/src/umath/extobj.c:
```diff
@@ -0,0 +1,317 @@
+
+#include "ufunc_object.h"  /* for npy_um_str_pyvals_name */
+
+#if USE_USE_DEFAULTS==1
```
Spaces around operators.

eric-wieser replied:
Again, this is verbatim from the other file.

charris commented on numpy/core/src/umath/extobj.c:
```diff
@@ -0,0 +1,317 @@
+    int res;
+
+    PyUFunc_NUM_NODEFAULTS += 1;
+    res = PyUFunc_GetPyValues("test", &bufsize, &errmask, &errobj);
```
"test"?

eric-wieser replied:
Beats me - again, just a maintenance change, moving this from the other file

charris commented on numpy/core/src/umath/extobj.c:
```diff
@@ -0,0 +1,317 @@
+        return 0;
+    }
+
+    NPY_ALLOW_C_API;
```
Does this buy anything? A small speedup in errorhandling, which should be rare, doesn't seem worth it.

charris replied:
NVM

eric-wieser commented:
Quoting the first commit:

> Apart from the includes, everything here is just a direct move - there are no code changes

I'd rather not open the can of poorly-documented worms that is how `extobj` actually works

eric-wieser commented:
@charris: Any other remarks here?

charris commented:
Nope, might as well put it in. Thanks Eric.

rgommers commented:
Please do not backport this - no need, and it will show warnings in the scipy test suite (and other packages probably) for released versions of scipy for which the test suite should be clean.
